### DISCUSSION TITLE
Rerandomizing problems with max attempts set

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/webwork2#2224] [General]
It will of course cause issues for users if your course is set to rerandomize problems after say 3 attempts, and also a homework set has problems capped at say 4 attempts. This came up locally with an instructor who turned on the rerandomization feature, but had inherited someone else's set definition files with the number of attempts capped.

It occurred to me that it might sort of make sense to do better. (Or at least different?) Like:

- reset a user's attempts count when a problem rerandomizes
- ignore the cap when rerandomization is enabled
- ignore rerandomization when there is a finite attempt cap
- throw a warning when you try to save the randomization config or set attempt caps (including importing a set) that causes this kind of clash

Or maybe some users like to rerandomize after 3 and cap at 6 attempts? In that case maybe if rerandomization is on, the cap attempt should actually be about how many times you are allowed to rerandomize, instead of counting each actual attempt? I like this last idea, but it would be a behavior change.

**Comment by drgrice1** [comment]
I like your last idea of the max attempts being the number of times you are allowed to randomize. I am not sure that it is that significant of a behavior change.

**Comment by Alex-Jordan** [comment]
After thinking about it more, I'm not sure any change would be good.

Returning to the scenario that started this, I guess it's not absurd that an instructor would *want* that configuration. You get to try one version three times, and if you still have not succeeded, now you get one last chance but it has to be with a brand new version. Not my taste as an instructor, but not as absurd as I first thought. So maybe the way things are is best.

**Comment by somiaj** [comment]
This could be a configuration item for the re-randomization. I do like your original thought of the max attempts be the number of times the problem can be re-randomized. But I do also see the use case that you described. So we could add an option that the instructor can set either behavior: The max attempts is total number of attempts, or the number of times a problem can be randomized again.

**Comment by dlglin** [comment]
I have used this in roughly the way that @Alex-Jordan describes: I have a checkbox problem where I give the students 2 attempts before re-randomizing, but give them 6 attempts overall, so they get a total of three versions.

I think that changing max attempts to be the number of times the student can re-randomize would be more confusing.  To me that seems inconsistent with the existing behaviour.
